src/models/feature.rs

```diff
@@ -7,11 +7,36 @@ pub type FeaturesMap = BTreeMap<String, Vec<String>>;
 ///
 /// See <https://rust-lang.github.io/rfcs/3143-cargo-weak-namespaced-features.html>.
 pub fn split_features(features: FeaturesMap) -> (FeaturesMap, FeaturesMap) {
-    features.into_iter().partition(|(_k, vals)| {
-        !vals
-            .iter()
-            .any(|v| v.starts_with("dep:") || v.contains("?/"))
-    })
+    // First, we partition the features into two groups: those that use the new
+    // features syntax (`features2`) and those that don't (`features`).
+    let (mut features, mut features2) =
+        features
+            .into_iter()
+            .partition::<FeaturesMap, _>(|(_k, vals)| {
+                !vals
+                    .iter()
+                    .any(|v| v.starts_with("dep:") || v.contains("?/"))
+            });
+
+    // Then, we recursively move features from `features` to `features2` if they
+    // depend on features in `features2`.
+    loop {
+        let split = features
+            .into_iter()
+            .partition::<FeaturesMap, _>(|(_k, vals)| {
+                !vals.iter().any(|v| features2.contains_key(v))
+            });
+
+        features = split.0;
+
+        if !split.1.is_empty() {
+            features2.extend(split.1);
+        } else {
+            break;
+        }
+    }
```

Comment:
yes, I'm a little worried about the infinite loop, but are there actually cases where this could actually get into infinite behavior? I don't think so, but a second(/third/fourth/...) pair of eyes won't hurt...

if we're worried that we may have missed something we could potentially also limit the number of iterations to a somewhat sane number roughly matching our features limit :)

Reply:
some additional context: this function is (currently) only used by the "sync to index" background jobs. in other words: this should not be able to freeze our API server, and if the background worker freezes up we will quickly see it in our monitoring.

Reply:
I've changed the code now to limit the number of "recursion" steps to 100 to play it safe...

Reply:
Do we want to add some logging when the recursion limit is reached? That way, we could examine what actually happened later.

Reply:
good call! implemented by https://github.com/rust-lang/crates.io/pull/9580

Reply:
I am also interested when the limit is reached. Where could I subscribe it? :)

Reply:
unfortunately you'd need access to the crates.io logs 😅